sandbox/apparmor/notify/listener/listener.go

```diff
@@ -0,0 +1,365 @@
+			// Must have been by the Close() function.
+			return ErrListenerAlreadyClosed
+		}
+		return err
+	}
+	for _, event := range events {
+		if event.Fd != int(l.notifyFile.Fd()) {
+			logger.Debugf("unexpected event from fd %v (%v)", event.Fd, event.Readiness)
+			continue
+		}
+		if event.Readiness&epoll.Readable == 0 {
+			continue
+		}
+		// Prepare a receive buffer for incoming request. The buffer is of the
+		// maximum allowed size and will contain one kernel request upon return.
+		// Note that the actually occupied buffer is indicated by the Length field
+		// in the header.
+		ioctlBuf := notify.NewIoctlRequestBuffer()
+		buf, err := notifyIoctl(l.notifyFile.Fd(), notify.APPARMOR_NOTIF_RECV, ioctlBuf)
+		if err != nil {
+			if l.poll.IsClosed() {
+				return ErrListenerAlreadyClosed
+			}
+			return err
+		}
+		if err := l.decodeAndDispatchRequest(buf); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (l *Listener) decodeAndDispatchRequest(buf []byte) error {
+	var nmsg notify.MsgNotification
+	if err := nmsg.UnmarshalBinary(buf); err != nil {
+		return err
+	}
+	// What kind of notification message did we get?
+	if nmsg.NotificationType != notify.APPARMOR_NOTIF_OP {
+		return fmt.Errorf("unsupported notification type: %v", nmsg.NotificationType)
+	}
+	var omsg notify.MsgNotificationOp
+	if err := omsg.UnmarshalBinary(buf); err != nil {
+		return err
+	}
+	// What kind of operation notification did we get?
+	switch omsg.Class {
+	case notify.AA_CLASS_FILE:
+		return l.handleRequestAaClassFile(buf)
+	}
+	return fmt.Errorf("unsupported mediation class: %v", omsg.Class)
+}
+
+func (l *Listener) handleRequestAaClassFile(buf []byte) error {
+	var fmsg notify.MsgNotificationFile
+	if err := fmsg.UnmarshalBinary(buf); err != nil {
+		return err
+	}
+	logger.Noticef("Received access request from the kernel: %+v", fmsg)
+	req, err := newRequest(&fmsg)
+	if err != nil {
+		return err
+	}
+	l.reqs <- req
+	l.tomb.Go(func() error {
+		return l.waitAndRespondAaClassFile(req, &fmsg)
+	})
+	return nil
+}
+
+func (l *Listener) waitAndRespondAaClassFile(req *Request, msg *notify.MsgNotificationFile) error {
+	resp := notify.ResponseForRequest(&msg.MsgNotification)
+	resp.MsgNotification.Error = 0 // ignored in responses
+	resp.MsgNotification.NoCache = 1
+	var allow bool
+	select {
+	case reply := <-req.YesNo:
+		allow = reply
+	case <-l.tomb.Dying():
+		logger.Noticef("tomb dying while waiting for reply to req: %+v; sending deny response", req)
+		allow = false
+	}
+	if allow {
+		// allow permissions which kernel initially denied, along with those which were already allowed
+		resp.Allow = msg.Allow | msg.Deny
+		resp.Deny = 0
+		resp.Error = 0
+	} else {
+		resp.Allow = msg.Allow
+		resp.Deny = msg.Deny
+		resp.Error = msg.Error
+		// msg.Error is field from MsgNotificationResponse, and is unused.
+		// msg.MsgNotification.Error is also ignored in responses.
+	}
+	logger.Noticef("Sending access response back to the kernel: %+v", resp)
```

Comment:
This seems like it might be more appropriate as a debug, might be too verbose to be in the hot path